### ISSUE TITLE
XGBoost Model Performance: B200 V A100/H100

### ISSUE THREAD

**Issue by ben-b-bll** [dmlc/xgboost#11608]
Hello team,

XGB Version: 3.0.3
CUDA Version: 12.8
NVIDIA Version: 570.172.08

XGBoost is an important part my workflow and I am now getting a relatively severe downgraded performance when running a XGBRegressor model on a B200 DGX (8x) versus A100 or H100s DGXs (8x).

I have tried running the model on various RapidsAI docker images which stay consistent across all machines. 

The train / test data and XGBoost params are consistent, the differences in test results are statistically significant. 

The data has been pre-processed and is a constant (being loaded in on each machine). The A100 == H100, but the B200<<A100/H100.

I've tried loading the data via many methods pandas, polars, cudf, into a QuantileDMatrix etc and the performance of the regression model is always degraded. I've noticed the same for classification models also.

Training is on single card for the above, standard xgb device=f'cuda:{n}', nothing multi-gpu going on via dask or similar. 

Is this is known issue with something that is being handled differently on the Blackwell architecture? 

Thank you team!

**Comment by trivialfis** [comment]
Hi, did you mean model accuracy? Could you please share the shape and sparsity of the dataset? If you have custom hyper-parameters, would be great to share as well.

**Comment by ben-b-bll** [comment]
Apologies, yes the model accuracy (however we want to measure it - RMSE for Reg, logloss for Class in these cases).

Speed performance improvements are as expected.... B200>H100>A100.

It's not a massively sparse dataset. 

<img width="822" height="239" alt="Image" src="https://github.com/user-attachments/assets/a73fe682-78ab-426c-b539-632f0d51a2c4" />

Nothing really custom XGB param wise, I have tried changing these and even building vanilla and still notice the same: 

```
model = xgb.XGBRegressor(random_state=777, booster='gbtree', tree_method='hist',early_stopping_rounds = 300,
                          n_estimators=20000, verbosity = 0,eta= 0.01,min_child_weight=400, max_depth=12, device='cuda:4', gamma=0.1, alpha=0.1, colsample_bytree= 0.7,subsample= 0.75)

model.fit(X_train[min_features], y_train[target], eval_set=[(X_test[min_features],y_test[target])], verbose=100)

```

**Comment by trivialfis** [comment]
Thank you for sharing, I will try to reproduce it by using a synthetic dataset first.

**Comment by trivialfis** [comment]
I have run some tests with the master branch, but I haven't tested the 3.0.3 release yet. I don't recall any related fix in the master branch that wasn't backported to 3.0.x.

Thus far, using the following synthetic data and parameters, I got bit-to-bit identical models between H200 and B200:

```
--n_samples_per_batch=2068509 --n_features=100 --n_batches=1 --device=cuda --sparsity=0.126
--eta=0.01 --min_child_weight=400 --max_depth=12 --colsample_bytree=0.7 --subsample=0.75 --gamma=0.1 --alpha=0.1  --n_rounds=20000 --task=qdm --seed=777
```

**Comment by trivialfis** [comment]
Could you please share a reproducible example?

**Comment by ben-b-bll** [comment]
Extremely strange.

I've upgraded CUDA to 13.0 and associated toolkits on the DGX with the B200s:

- Still getting noticeably downgraded GPU performance (identical to what is reported to the above)
- If I run on the CPU on the B200 I get no downgraded performance - statistically similar results to the H100s machine.

I have tried using multiple versions of XGBoost - will try some older now.

Catboost, utilising the GPU, yields almost identical results across both machines.

**Comment by trivialfis** [comment]
Could you please share a reproducible example, maybe with synthetic datasets?

**Comment by ben-b-bll** [comment]
Update on the B200 (Blackwell) v H100&A100 DGXs:

xgboost==2.1.1 - yields statistically similar outputs now.

xgboost==3.0.0 - has degraded performance.

xgboost==2.1.4 - has degraded performance.

xgboost==2.1.3 - yields statistically similar outputs now.

Will share reproducible example.

**Comment by trivialfis** [comment]
> xgboost==3.0.0 - has degraded performance.

In the meantime, did you observe the effect using 3.0.0, 3.0.3, or both?

**Comment by ben-b-bll** [comment]
All xgboost releases 2.1.4 onwards seem to cause degraded performance (in terms of prediction quality) on the B200s v the H100/A100s. 

Host machine tested with CUDA 12.8 and CUDA 13.0 and associated toolkits. 

Please let me know if you need any more information.

<img width="911" height="46" alt="Image" src="https://github.com/user-attachments/assets/0ef5f011-eb89-433c-8c4c-968e01aec1ef" />

**Comment by trivialfis** [comment]
> Please let me know if you need any more information.

I don't have any guesses at the moment. Will dig deeper once there's a reproducible example

**Comment by ben-b-bll** [comment]
> > 
> 
> In the meantime, did you observe the effect using 3.0.0, 3.0.3, or both?

I've tried every version 2.1.4 onwards and observed the same. 

Every previous version produces results as expected. 

Exact same machine, environment, container, dataset, params etc.

**Comment by ben-b-bll** [comment]
<img width="719" height="274" alt="Image" src="https://github.com/user-attachments/assets/a26613a1-cf64-4f18-a297-bc72fc742d75" />

Seems to be introduced in the version above.

Please find code example below, ran on DGX B200.

```
from sklearn.datasets import make_regression
from sklearn.model_selection import train_test_split
from sklearn.metrics import root_mean_squared_error
import xgboost as xgb

# Step 1: Create synthetic regression data
X, y = make_regression(n_samples=100_000, n_features=200, noise=0.2, random_state=42)

# Step 2: Split into train/test
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print(xgb.__version__)
model = xgb.XGBRegressor(
    tree_method="hist",  
    n_estimators=1000,
    max_depth=8,
    eta = 0.05,
     device='cuda:7',
    colsample_bytree= 0.7,
    subsample= 0.7,
    min_child_weight=100, 
    gamma = 0.01,
    alpha = 0.01,
    early_stopping_rounds=100
)

model.fit(X_train, y_train, eval_set=[(X_test,y_test)], verbose=100)

y_pred = model.predict(X_test)
rmse = root_mean_squared_error(y_test, y_pred)
print(f"RMSE: {rmse:.2f}")
```

`3.0.2
[0]	validation_0-rmse:196.90496
[100]	validation_0-rmse:49.76079
[200]	validation_0-rmse:36.00881
[300]	validation_0-rmse:33.26109
[400]	validation_0-rmse:31.94906
[500]	validation_0-rmse:31.13464
[600]	validation_0-rmse:30.24052
[700]	validation_0-rmse:29.55064
[800]	validation_0-rmse:29.10030
[900]	validation_0-rmse:28.82976
[999]	validation_0-rmse:28.69853
RMSE: 28.70`

```
!pip install xgboost==2.1.1

Collecting xgboost==2.1.1
  Downloading xgboost-2.1.1-py3-none-manylinux_2_28_x86_64.whl.metadata (2.1 kB)
Requirement already satisfied: numpy in /opt/conda/lib/python3.13/site-packages (from xgboost==2.1.1) (2.2.6)
Collecting nvidia-nccl-cu12 (from xgboost==2.1.1)
  Downloading nvidia_nccl_cu12-2.27.7-py3-none-manylinux2014_x86_64.manylinux_2_17_x86_64.whl.metadata (2.0 kB)
Requirement already satisfied: scipy in /opt/conda/lib/python3.13/site-packages (from xgboost==2.1.1) (1.16.0)
Downloading xgboost-2.1.1-py3-none-manylinux_2_28_x86_64.whl (153.9 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 153.9/153.9 MB 11.8 MB/s eta 0:00:0000:0100:01
Downloading nvidia_nccl_cu12-2.27.7-py3-none-manylinux2014_x86_64.manylinux_2_17_x86_64.whl (322.5 MB)
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 322.5/322.5 MB 11.5 MB/s eta 0:00:0000:0100:01
Installing collected packages: nvidia-nccl-cu12, xgboost
  Attempting uninstall: xgboost━━━━━━━━━━━━━━━━━ 0/2 [nvidia-nccl-cu12]
    Found existing installation: xgboost 3.0.20m 0/2 [nvidia-nccl-cu12]
    Uninstalling xgboost-3.0.2:━━━━━━━━━━━━━ 0/2 [nvidia-nccl-cu12]
      Successfully uninstalled xgboost-3.0.2 0/2 [nvidia-nccl-cu12]
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2/2 [xgboost]m1/2 [xgboost]
Successfully installed nvidia-nccl-cu12-2.27.7 xgboost-2.1.1
```

`2.1.1
[0]	validation_0-rmse:199.93716
[100]	validation_0-rmse:35.15103
[200]	validation_0-rmse:18.93392
[300]	validation_0-rmse:16.74262
[400]	validation_0-rmse:16.38201
[500]	validation_0-rmse:16.28234
[600]	validation_0-rmse:16.24858
[700]	validation_0-rmse:16.24422
[800]	validation_0-rmse:16.23274
[900]	validation_0-rmse:16.22462
[999]	validation_0-rmse:16.21013
RMSE: 16.21`

**Comment by trivialfis** [comment]
Thank you, this is really helpful. My previous test showed that https://github.com/dmlc/xgboost/pull/11542 should fix the discrepancy on blackwell, but it turns out https://github.com/dmlc/xgboost/pull/11540 might be necessary as well. Not sure what the compiler is doing under the hood, still investigating.

In the meantime, you can test the XGBoost nightly build: https://xgboost.readthedocs.io/en/stable/install.html#id1 The latest commit should contain everything needed.

**Comment by ben-b-bll** [comment]
> Thank you, this is really helpful. My previous test showed that [#11542](https://github.com/dmlc/xgboost/pull/11542) should fix the discrepancy on blackwell, but it turns out [#11540](https://github.com/dmlc/xgboost/pull/11540) might be necessary as well. Not sure what the compiler is doing under the hood, still investigating.
> 
> In the meantime, you can test the XGBoost nightly build: https://xgboost.readthedocs.io/en/stable/install.html#id1 The latest commit should contain everything needed.

`!pip install https://s3-us-west-2.amazonaws.com/xgboost-nightly-builds/master/f0d6a256119fb38f8da8805703119c17cd279c66/xgboost-3.1.0.dev0-py3-none-manylinux_2_28_x86_64.whl`

`3.1.0-dev
[0]	validation_0-rmse:196.38169
[100]	validation_0-rmse:34.83708
[200]	validation_0-rmse:18.72093
[300]	validation_0-rmse:16.48741
[400]	validation_0-rmse:16.12837
[500]	validation_0-rmse:16.05189
[600]	validation_0-rmse:16.01562
[700]	validation_0-rmse:15.99390
[800]	validation_0-rmse:15.99656
[835]	validation_0-rmse:15.99891
RMSE: 15.99`

Seems to fix the issue.... 

Can confirm when I downgrade back to 3.0.3 I get the old output as previously shared.

No problems using that dev build moving forward is there? Assume these changes to be pushed to 3.0.4 at some point? Thanks!

**Comment by trivialfis** [comment]
Let me do some more investigation.

**Comment by trivialfis** [comment]
I will get this fix in https://github.com/dmlc/xgboost/pull/11616 .

**Comment by trivialfis** [comment]
> No problems using that dev build moving forward is there? Assume these changes to be pushed to 3.0.4 at some point? Thanks!

You can use it with commits after https://github.com/dmlc/xgboost/pull/11616 . (749eff7add89f0cf05dcb44799735f8c7d3de482) . I'm planning for the 3.1 release now and want to use smaller releases than XGBoost used to have. I don't have an eta yet.

**Comment by trivialfis** [comment]
Feel free to reopen if there are further issues.

**Comment by ben-b-bll** [comment]
Awesome, thanks for your rapid turnaround. 

Can confirm that works as expected now on B200 DGX post https://github.com/dmlc/xgboost/pull/11616

**Comment by trivialfis** [comment]
Just made a patch release that includes the fix.
